specs/hierarchical_consensus.md

```diff
@@ -0,0 +1,1082 @@
+    // the consensus protocol (e.g. minimum number of validators).
+    // These parameters are consensus-specific.
+    ConsensusParams *hierarchical.ConsensusParams
+    // Minimum amount of stake required from participant 
+    // to become a validator of the subnet.
+    MinValidatorStake   abi.TokenAmount
+    // Checkpointing period used in the subnet. It determines
+    // how often the subnet propagates and commits checkpoints in its
+    // parent
+    CheckPeriod     abi.ChainEpoch
+}
+```
+> TODO: There is an on-going revision of `ConstructParams` to include a way to provide arbitrary input arguments to a subnet.
+
+## Subnet Coordinator Actor (SCA)
+The main entity responsible for handling all the lifecycle of child subnets in a specific chain is the Subnet Coordinator Actor (`SCA`). The `SCA` is a builtin-actor that exposes the interface for subnets to interact with the hierarchical consensus protocol. This actor includes all the available functionalities related to subnets and their management. It also enforces all the security requirements, fund management, and cryptoeconomics of the hierarchical consensus, as Subnet Actors are user-defined and can't be (fully) trusted. The `SCA` has a reserved address ID `f064`.
+
+The MVP implementation of this builtin-actor can be found [here](https://github.com/adlrocha/builtin-actors/tree/master/actors/hierarchical_sca). The `SCA` exposes the following functions:
+
+
+##### SCA Functions
+> Functions of the SCA. These functions are triggered when a `message` is sent for their corresponding `methodNum`.
+```go
+type SCA interface{
+    // Initializes the state of the SCA and sets
+    // all its initial parameters.
+    // (methodNum = 1)
+    Constructor(ConstructParams)
+    
+    // Register expects as source the address of the subnet actor of
+    // the subnet that wants to be registered. Its
+    // `value` an amount of collateral over the `CollateralThreshold`.
+    // This functions activates the subnet and from there on other
+    // subnets in the system are allowed to interact with it and the
+    // subnet can start commtting its checkpoints.
+    // (methodNum = 2)
+    Register()
+    
+    // AddCollateral expects as source the address of the subnet actor
+    // for which the collateral wants to be added.  Its `value` should
+    // include the amount of collateral to be added for the subnet.
+    // (methodNum = 3)
+    AddCollateral()
+    
+    // ReleaseCollateral expects as source the address of the subnet actor
+    // for which the collateral should be released. It triggers a message
+    // to the subnet actor returning the corresponding collateral. 
+    // (methodNum = 4)
+    ReleaseCollateral(value abi.TokenAmount)
+    
+    // Kill expects as source the address of the subnet actor to be killed.
+    // This function can only be executed if no collateral or circulating
+    // supply is left for the subnet (i.e. balance = 0).
```

Comment:
Curious: this check has not been mentioned in the subnet actor interface, it just says "sanity checks". From what I saw, first there's the actor state transition in a `transaction`, and then there's the `send`. What if somebody implements an actor which doesn't adhere to this rule, kills their subnet actor, and then sends the message to the SCA, but the SCA refuses to kill it?

Reply:
Great question. If I am not mistaken, in the current implementation of the VM side-effects are atomic, meaning that when `Kill()` is called in the subnet actor, the set of checks are run, and if they pass a side-effect message is sent to `SCA::Kill`. If the latter fails, the whole message will fail, so there's no way for the subnet actor kill to succeed without the SCA Kill succeeding. 

That being said, I don't know if this will change for FVM-M2 (I'll try to find where this is being discussed). I agree that if this is the case we'll need to be careful and revisit how we kill subnets.

Reply:
Interesting. Yeah the semantics seem to be left the the actor coder, because `Runtime` is way too powerful. It is handed to the actor as-is, and you are expected to use `transaction` and not call `send` while doing so. But you are not forced to do so - you can retrieve your state with `runtime.state()`, do your thing and use `send`, then use `transaction` only to set the state to the value you updated earlier. 

I agree that it probably won't commit if send fails, but I also didn't understand what `send` does, whether it executes immediately, or just lines up a message to be done later. Looked like the latter. 

I suppose using transactions and not sending messages in the middle prevents some reentrancy bugs, but because it's not enforced and it looks like you have to have discipline to stick to it and code it right, it would be good to hear the full story of where this is heading.